pex/build_system/pep_518.py

```diff
@@ -0,0 +1,183 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import absolute_import
+
+import json
+import os.path
+import pkgutil
+import subprocess
+from subprocess import CalledProcessError
+from textwrap import dedent
+
+from pex.common import atomic_directory
+from pex.dist_metadata import Distribution
+from pex.pex import PEX
+from pex.pex_bootstrapper import ensure_venv
+from pex.pex_builder import PEXBuilder
+from pex.resolve.lockfile import json_codec
+from pex.resolve.resolvers import Resolver
+from pex.result import Error
+from pex.third_party import isolated
+from pex.tracer import TRACER
+from pex.typing import TYPE_CHECKING
+from pex.variables import ENV
+from pex.venv.bin_path import BinPath
+
+if TYPE_CHECKING:
+    from typing import Iterable, Mapping, Optional, Text, Tuple, Union
+
+    import attr  # vendor:skip
+else:
+    from pex.third_party import attr
+
+
+def _execute_with_toml(
+    resolver,  # type: Resolver
+    script_contents,  # type: str
+):
+    # type: (...) -> Union[Text, Error]
+    target_dir = os.path.join(ENV.PEX_ROOT, "tools", "toml.pex", isolated().pex_hash)
+    with TRACER.timed("Building toml parsing PEX"):
+        with atomic_directory(target_dir=target_dir, exclusive=True) as atomic_dir:
+            if not atomic_dir.is_finalized():
+                lock_file_contents = pkgutil.get_data(__name__, "toml-lock.json")
+                assert lock_file_contents is not None, (
+                    "The sibling resource toml-lock.json of {} should always be present in a "
+                    "Pex distribution or source tree.".format(__name__)
+                )
+                contents = lock_file_contents.decode("utf-8")
+                source = os.path.join(os.path.dirname(__file__), "toml-lock.json")
+                lock = json_codec.loads(lockfile_contents=contents, source=source)
+
+                # Our toml lock was created against the default public PyPI index; so we can only
+                # use the lock if the resolve is also setup to use the default repos configuration
+                # as well.
+                if resolver.is_default_repos():
```

Comment:
IIUC if the PipConfiguration differs from the default in ways that could still allow lockfile use then we won't use the lockfile. That's fine but the one case I wonder about is repos_configuration including standard PyPI but also adding private repos. But that seems like a bridge we can cross when we come to it.

Reply:
This is moot since I'll just vendor toml per above, but not using the lockfile even when we could've is only slower, not less correct, since the pinned version here is a dependency free dist. In other words, we just go through Pip hoops for it to eventually figure out there is only 1 project needed instead of directly fetching the locked dist URL. As such, instead of wading into somewhat titchy things - namely ensuring PyPI is the 1st index URL in the custom indexes list and double checking Pip uses indexes 1st than falls back to repos